### PR TITLE
contrib: Parse ELF directly for symbol and security checks

### PR DESCRIPTION
**Pull request by laanwj**
Instead of the ever-messier text parsing of the output of the readelf tool (which is clearly meant for human consumption not to be machine parseable), parse the ELF binaries directly.

Add a small dependency-less ELF parser specific to the checks.

This is slightly more secure, too, because it removes potential ambiguity due to misparsing and changes in the output format of `elfread`. It also allows for stricter and more specific ELF format checks in the future.

This removes the build-time dependency for `readelf`.

It passes the test-security-check for me locally, ~~though I haven't checked on all platforms~~. I've checked that this works on the cross-compile output for all ELF platforms supported by Bitcoin Core at the moment, as well as PPC64 LE and BE.

### CODE REVIEW

**Comment by fanquake** [comment]
Concept ACK 🧚

**Comment by jonatack** [comment]
Concept ACK

**Comment by laanwj** [comment]
I've ported the changes for PPC64 from #14066 in an extra commit.

File: contrib/devtools/symbol-check.py [hunk 1]
```diff
@@ -142,26 +145,15 @@ def close(self):
 
 def read_symbols(executable, imports=True) -> List[Tuple[str, str, str]]:
```
Comment (laanwj) [inline contrib/devtools/symbol-check.py:146]: I don't know much about python type specifications but this is wrong now, I think it is `Tuple[List[Tuple[str, str]], int]`.

**Comment by laanwj** [comment]
Added mypy annotations. This found a few (small) actual bugs, and might make review easier.

**Comment by theStack** [comment]
Concept ACK

**Comment by sipa** [comment]
Concept ACK.

ELF looks remarkably simple!

File: contrib/devtools/pixie.py [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+'''
+Compact, self-contained ELF implementation for bitcoin-core security checks.
+'''
+import struct
+import types
+from typing import Dict, List, Optional, Union, Tuple
+
+# you can find all these values in elf.h
+EI_NIDENT = 16
```
Comment (fanquake) [inline contrib/devtools/pixie.py:13]: note: Now that we require Python 3.6, if you want you can also use [variable annotations](https://www.python.org/dev/peps/pep-0526/):
```suggestion
EI_NIDENT: int = 16
```

Comment (laanwj) [reply contrib/devtools/pixie.py:13]: Sure but, I added any annotations that `mypy` required of me even with various `--disallow-*` options. I'm not convinced that decorating unambiguous assignments of integers with `: int` adds anything but verbosity.

**Comment by laanwj** [comment]
> ELF looks remarkably simple!

The secret is not needing to resolve relocations :slightly_smiling_face: That is where most of the verbose platform specific trickery comes in.

**Comment by practicalswift** [comment]
Concept ACK: removing dependencies improves robustness

**Comment by laanwj** [comment]
It's not *really* the dependency that is the problem here. Sure, needing to call out to an utility less is nice, but `readelf` is part of binutils that we require for build. If it had a stable, predictable output format for machine parsing (like git's plumbing/porcelain distinction), this wouldn't be needed. The problem is basing security checks on an ambiguous human-friendly output format.

**Comment by luke-jr** [comment]
Concept ACK

**Comment by laanwj** [comment]
Would be nice to get #20476 in first, which adds a test that the symbol check can actually fail.

**Comment by laanwj** [comment]
Good to see that both guix and gitian builds passed for all platforms.

**Comment by laanwj** [comment]
Closing and reopening to re-trigger Travis with #20476 in.

**Comment by MarcoFalke** [comment]
cirrus won't re-trigger on open-close, but on a push only (e.g. rebase or other any other git push)

**Comment by laanwj** [comment]
*sigh* okay, added an empty commit to re-trgger CI.

**Comment by fanquake** [comment]
Just wanted  to note that this will fail the `RELRO` check if run against binaries older than 0.17.x, because they have the deprecated `DT_BIND_NOW` array tag:
```bash
/usr/local/opt/binutils/bin/readelf -dw ~/Downloads/bitcoin-0.16.3/bin/bitcoind | rg -i bind
...
 0x0000000000000009 (RELAENT)            24 (bytes)
 0x0000000000000018 (BIND_NOW)           
 0x000000006ffffffb (FLAGS_1)            Flags: NOW
```

rather than the `DF_BIND_NOW` flag:
```bash
/usr/local/opt/binutils/bin/readelf -dw ~/Downloads/bitcoin-0.17.2/bin/bitcoind | rg -i bind
...
 0x0000000000000009 (RELAENT)            24 (bytes)
 0x000000000000001e (FLAGS)              BIND_NOW
 0x000000006ffffffb (FLAGS_1)            Flags: NOW PIE
```

and our current script [checks for either](https://github.com/bitcoin/bitcoin/blob/25bc840e038f5cc3a1d70d6fbaf1dc2eb2c6e460/contrib/devtools/security-check.py#L123). Not that it's a blocker, just noticed while testing.

**Comment by laanwj** [comment]
Thanks for testing!

In contrast to passing, failing is safe in this case. I think it's fine to support only the 'new way' only.

**Comment by sipa** [comment]
Concept & approach ACK. I have tried the pixie module on a few binaries and seems to give reasonable results, but haven't reviewed the code in detail or verified against any ELF format reference.

**Comment by laanwj** [comment]
@sipa Thanks for testing and looking at it!

FWIW, verifying it against `elf.h` is probably easier than verifying it against the ELF spec as it follows a similar structure. But I hope the security and symbol-check tests are enough to give confidence that this has the same functionality as the `elfread` based implementation.
(apart from the deprecated `DT_BIND_NOW` check, apparently)

**Comment by laanwj** [comment]
I'm going to merge this so that #14066 can be rebased on top and we can introduce the POWER architecture binaries in the beginning of the 0.22 window instead of the last minute (or again miss it last minute).

**Pull request merged**